### DISCUSSION TITLE
Custom Nodes and Malt Nodes API?

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#240] [Q&A]
I am looking into creating custom nodes for Malt. I have started with trying to recreate the MixRGB node that comes with regular old shader nodes. The challenges I am facing are:

- [x] Get all the shader code 
- [x] Registering and unregistering a custom node that integrates into the node tree evaluation pipeline without it requiring extra integration code
- [x] create a UI (enum property) that switches between the shader code 

I already got the code and know how to code custom nodes in general. If I am not mistaken, custom nodes should not require any extra integration. The one thing that matters is the `MaltFunctionNode.get_source_code( )` function, right? Also if I understand it correctly, the nodes do NOT store their own code but access a library compiled out of all sources of shader functions (built-in, global plugins, local plugins, local libraries). My question would be **what is the best approach to creating custom nodes for malt?** Override the `MaltFunctionNode.get_source_code( )` function to return a string stored INSIDE the node or having the custom node come in a plugin that contains both the custom node as well as the shader code so that the custom node only ever references a function name? 
Btw, chaning the attribute `MaltFunctionNode.function_type` triggers the Node to update but not the node tree. I know you are not expecting anyone to fiddle around there but is that something that should be added when updating the node function type?

**Comment by pragma37** [comment]
This is a complex question.

First, a pedantic note in case other people arrive here looking for how to add custom nodes in Malt.
**The easiest way to add custom nodes to Malt is by simply adding a custom library**, either directly in a node tree:
https://youtu.be/tE99jgCCcNE?t=705
Or through a plugin:
https://github.com/bnpr/Malt/tree/Development/plugins/PluginExample

This question is about adding nodes to BlenderMalt, using the Blender API.

Any custom BlenderMalt node should inherit from both [bpy.types.Node](https://docs.blender.org/api/current/bpy.types.Node.html) and [MaltNode](https://github.com/bnpr/Malt/blob/Development/BlenderMalt/MaltNodes/MaltNode.py).
At the very least, it should probably have to implement malt_init, malt_setup, get_source_socket_reference and get_source_code.
The simplest reference is the [MaltArrayIndexNode](https://github.com/bnpr/Malt/blob/Development/BlenderMalt/MaltNodes/Nodes/MaltArrayIndexNode.py).

For getting it into the SHIFT+A menu you can [append](https://docs.blender.org/api/current/bpy.types.Menu.html#extending-menus) it into bpy.types.NODE_MT_add. Or any of the [Malt specific submenus](https://github.com/bnpr/Malt/blob/6e94c5f6cac96f5e80fe9a995a01286396c965aF3/BlenderMalt/MaltNodes/MaltNodeTree.py#L514).

The top level function where the source code is generated is [MaltTree.get_generated_source()](https://github.com/bnpr/Malt/blob/Development/BlenderMalt/MaltNodes/MaltNodeTree.py#L125).
It generates the code by simply concatenating in order the strings returned by nodes in 2 different scopes:
- The global scope (Node.get_source_global_parameters). Everything that should be in the global GLSL scope (uniforms, global variables...) should go here.
- The local scope (Node.get_source_code). This is the body of the function that is being implemented (PRE_PASS_PIXEL_SHADER, MAIN_PASS_PIXEL_SHADER...).

So in the case of a MixRGB node, you could either add the functions to a plugin library (or declare them on the global scope) and then call them in get_source_code. Or simply inlining the code in get_source_code.

> Btw, chaning the attribute MaltFunctionNode.function_type triggers the Node to update but not the node tree. I know you are not expecting anyone to fiddle around there but is that something that should be added when updating the node function type?

This is an internal API. You're not supposed to touch it from the outside.
If you do it's up to you to make sure it behaves correctly.
Simply updating the node tree for every node update would be very slow.

**(Edit)**
Blender addon example:
*This node takes 2 float parameters (a and b) and adds them (result = a+b).*
```python
bl_info = {
    "name" : "Malt Node Addon Example",
    "author" : "Miguel Pozo",
    "description" : "",
    "blender" : (3, 0, 0),
    "category" : "Script"
}

import bpy    
from BlenderMalt.MaltNodes.MaltNode import MaltNode

class MaltAddonExampleNode(bpy.types.Node, MaltNode):
    
    bl_label = "Addon Example Node"

    def malt_init(self):
        self.setup()
        
    def malt_setup(self):
        if self.first_setup:
            self.name = 'Addon Example'
        inputs = {
            'a' : {'type': 'float'},
            'b' : {'type': 'float'},
        }
        outputs = {
            'result' : {'type': 'float'},
        }
        self.setup_sockets(inputs, outputs)

    def get_source_socket_reference(self, socket):
        transpiler = self.id_data.get_transpiler()
        return transpiler.parameter_reference(self.get_source_name(), socket.name, 'out' if socket.is_output else 'in')
    
    def get_source_code(self, transpiler):
        a = self.inputs['a']
        b = self.inputs['b']
        result = self.outputs['result']
        initialization = '{} + {}'.format(a.get_source_initialization(), b.get_source_initialization())
        source = transpiler.declaration(result.data_type, result.array_size, result.get_source_reference(), initialization)
        print(source)
        return source

def add_node_ui(self, context):
    sd = context.space_data
    if sd.tree_type == 'MaltTree' and sd.node_tree and sd.node_tree.graph_type != '':
        operator = self.layout.operator("node.add_node", text='Addon Node Example')
        operator.type = 'MaltAddonExampleNode'
        operator.use_transform = True

def register():
    bpy.utils.register_class(MaltAddonExampleNode)
    bpy.types.NODE_MT_add.append(add_node_ui)
    
def unregister():
    bpy.types.NODE_MT_add.remove(add_node_ui)
    bpy.utils.unregister_class(MaltAddonExampleNode)
```
---
(2º Edit)
A second example for more complex functions. It inlines the alpha_blend code from Shaders/Common/Color.glsl.
Note how all the code except the result declaration is inside a local scope block to avoid variable name collisions.
The return statements have been replaced by a result assignment.
It also checks whether the node tree language is GLSL since this obviously won't compile in Python graphs like Render and Render Layer.
```python
bl_info = {
    "name" : "Malt Node Addon Example",
    "author" : "Miguel Pozo",
    "description" : "",
    "blender" : (3, 0, 0),
    "category" : "Script"
}

import bpy    
from BlenderMalt.MaltNodes.MaltNode import MaltNode

class MaltAddonExampleNode(bpy.types.Node, MaltNode):
    
    bl_label = "Addon Example Node"

    def malt_init(self):
        self.setup()
        
    def malt_setup(self):
        if self.first_setup:
            self.name = 'Addon Example'
        inputs = {
            'base' : {'type': 'vec4'},
            'blend' : {'type': 'vec4'},
        }
        outputs = {
            'result' : {'type': 'vec4'},
        }
        self.setup_sockets(inputs, outputs)

    def get_source_socket_reference(self, socket):
        transpiler = self.id_data.get_transpiler()
        return transpiler.parameter_reference(self.get_source_name(), socket.name, 'out' if socket.is_output else 'in')
    
    def get_source_code(self, transpiler):
        base = self.inputs['base']
        blend = self.inputs['blend']
        result = self.outputs['result']
        source = '''
        vec4 {result};
        {{
            vec4 base = {base};
            vec4 blend = {blend};
            vec4 result;
            
            if(blend.a <= 0)
            {{
                result = base;
            }}
            else if(blend.a >= 1)
            {{
                result = blend;
            }}
            else
            {{
                float alpha = blend.a + base.a * (1.0 - blend.a);
                result = (blend * blend.a + base * base.a * (1.0 - blend.a)) / alpha;
                result.a = alpha;
            }}
            
            {result} = result;
        }}
        '''.format(
            result=result.get_source_reference(),
            base=base.get_source_initialization(),
            blend=blend.get_source_initialization()
        )
        return source

def add_node_ui(self, context):
    sd = context.space_data
    if sd.tree_type == 'MaltTree' and sd.node_tree and sd.node_tree.graph_type != '' and sd.node_tree.get_source_language() == 'GLSL':
        operator = self.layout.operator("node.add_node", text='Addon Node Example')
        operator.type = 'MaltAddonExampleNode'
        operator.use_transform = True

def register():
    bpy.utils.register_class(MaltAddonExampleNode)
    bpy.types.NODE_MT_add.append(add_node_ui)
    
def unregister():
    bpy.types.NODE_MT_add.remove(add_node_ui)
    bpy.utils.unregister_class(MaltAddonExampleNode)
```

**Reply by pragma37** [reply]
BTW, I just realized that every time I touched those markdown tick boxes to see how they worked I was actually editing your post. 🤦‍♂️
Sorry!

**Reply by Kolupsy** [reply]
ah no worries. I just noticed that there were checkboxes and thought 'oh fancy ima try those'. I didnt get notified so didnt even realize 👀 
I wish I had understood what your reply meant. I definitely know how to make custom nodes but I will have to see how I override / define these mandatory functions. I may ask for a simple example in the near future 😅

**Reply by Kolupsy** [reply]
Ok here we go. First of I would like to know why the blender console is always empty when running malt. I guess the stdout gets redirected somewhere? It would be interesting to know where so that I can see whats going on. The other question is How do I integrate a cusom node? I would like to register is in the `__init__.py` file of my plugin if possible to keep everything nice and self-contained. But the default init file that comes with the example plugin only has a `register_graph_libraries` funcion:

```py
import os
from Malt.PipelinePlugin import PipelinePlugin, isinstance_str

class PluginExample(PipelinePlugin):

    @classmethod
    def poll_pipeline(self, pipeline):
        return isinstance_str(pipeline, 'NPR_Pipeline')
    
    @classmethod
    def register_graph_libraries(self, graphs):
        library_path = os.path.join(os.path.dirname(__file__), 'Shaders')
        for graph in graphs.values():
            if graph.language == 'GLSL':
                graph.add_library(library_path)

PLUGIN = PluginExample
```

**Reply by pragma37** [reply]
The stdout is overridden ![here](https://github.com/bnpr/Malt/blob/Development/Bridge/Client_API.py#L35).
However, the Blender console should still work. It certainly does for me on every system I tested.

The Malt plugins system is only for Malt plugins, you can't use the Blender API from there (if that's what you were planning).
And nope, GLSL functions have to go into their own files.

**Reply by Kolupsy** [reply]
Ok the console issue was on my end. it works now. 
about the malt plugin system: what do you mean 'i cant use the Blender API from there'? Is that a technical limitation or something that is not implemented yet? Because it would be super super useful if those `__init__.py` files could have register and unregister functions that are being called on `bpy.ops.wm.malt_reload_pipeline( )`. This would allow Malt plugins to be a lot more. If I was to create a plugin for malt that includes shader code and python nodes (MixRGB example) I would need to install the malt plugin and the blender plugin separately. Another example would be that you could integrate python into a malt plugin by for example setting up dafault nodes for every new malt tree or something like that. 
If thats not possible or too nieche of a problem to tackle then thats fine.

**Reply by pragma37** [reply]
> what do you mean 'i cant use the Blender API from there'? Is that a technical limitation or something that is not implemented yet?

Malt runs on a separate process, you can't access the Blender API from Malt.
Even if it was possible, it would prevent Malt from supporting anything but Blender.

>  I would need to install the malt plugin and the blender plugin separately.

Nope, BlenderMalt nodes only generate the glsl/python source code. You don't have to register anything as a Malt Plugin if you don't need it. (I've updated my answer with an example).

So, long story short. If you want to extend Malt, use a Malt Plugin. If you want to extend BlenderMalt, use a Blender Addon.

**Reply by Kolupsy** [reply]
Ah yes that makes sense. I keep forgetting that malt should conceptually be independant from blender. Ill try to come up with a concept of how to easily develop BlenderMalt plugins. Maybe inlining the shader node would be the best option then 🤔

**Comment by Kolupsy** [comment]
https://user-images.githubusercontent.com/88381426/153970320-6c64366b-8a1c-4678-95f1-f5293ef2ad76.mp4

I have got a mixRGB mockup working but its quite janky. It currently relies on having a Malt Plugin installed in addition to the BlenderMalt plugin that registers the custom node. That is something I would like to fix. Also I am having trouble working with the shader compilation stuff in malt. Right now I cannot really use proper formatted glsl together with python to create a custom node. Here is what I came up with:
```py
import bpy
from bpy.types import Node  
from bpy.props import *
from BlenderMalt.MaltNodes.MaltNode import MaltNode

blend_types = [
    ( 'MIX', 'Mix', 'mix_blend' ),
    ( 'ADD', 'Add', 'mix_add' ),
    ( 'SUBTRACT', 'Subtract', 'mix_subtract' ),
    ( 'MULTIPLY', 'Multiply', 'mix_multiply' ), 
]
def blend_types_items( ):
    return (( x[0], x[1], x[1] ) for x in blend_types )

def get_blend_function( blend_type ):
    return next( iter( x for x in blend_types if x[0] == blend_type ))[2]

class MaltCustomNodeMix( Node, MaltNode ):
    bl_label = "MixRGB Node"

    blend_type : EnumProperty( items = blend_types_items( ), name = 'Blend Type', update = lambda s, c: s.update_blend_type( ))
    def update_blend_type( self ):
        self.id_data.update( )

    def malt_init( self ):
        self.setup()
        
    def malt_setup(self):
        if self.first_setup:
            self.name = 'Mix'
        inputs = {
            'fac' : {'type': 'float'},
            'col1' : {'type': 'vec4'},
            'col2' : {'type': 'vec4'},
        }
        outputs = {
            'result' : {'type': 'vec4'},
        }
        self.setup_sockets( inputs, outputs )

    def get_source_socket_reference( self, socket ):
        transpiler = self.id_data.get_transpiler( )
        return transpiler.parameter_reference( self.get_source_name( ), socket.name, 'out' if socket.is_output else 'in' )

    def get_source_code(self, transpiler):
        fac = self.inputs['fac'].get_source_initialization( )
        col1 = self.inputs['col1'].get_source_initialization( )
        col2 = self.inputs['col2'].get_source_initialization( )
        result = self.outputs['result']
        res = result.get_source_reference( )
        f = get_blend_function( self.blend_type )
        initialization = '{0}( {1}, {2}, {3}, {4} );'.format( f, fac, col1, col2, res )
        source = f'{result.data_type} {res}; {initialization}'
        return source
    
    def draw_buttons( self, context, layout ):
        layout.prop( self, 'blend_type', text = 'Mix' )

    def add_node_ui( self, context ):
        sd = context.space_data
        if sd.tree_type == 'MaltTree' and sd.node_tree and sd.node_tree.graph_type != '':
            operator = self.layout.operator( "node.add_node", text = self.bl_label )
            operator.type = 'MaltCustomNodeMix'
            operator.use_transform = True

def get_classes( ):
    yield MaltCustomNodeMix
```
The part where i sloppily throw together the return string for `get_source_code( )` is what bothers me. This would not work with other functions or at least not with a great deal of manual converting. 
Ideally I just want to have the custom node point to a function in an attached glsl file, read that function and from that, set up the sockets and inlined shader code.

**Reply by pragma37** [reply]
Nice!

> Right now I cannot really use proper formatted glsl together with python to create a custom node. 

I don't get what this means. How "proper formatted glsl together with python" would look like?

> The part where i sloppily throw together the return string for get_source_code( ) is what bothers me. This would not work with other functions or at least not with a great deal of manual converting.

You should use the transpiler for handling this. Here's the whole API:
https://github.com/bnpr/Malt/blob/Development/Malt/SourceTranspiler.py

> Ideally I just want to have the custom node point to a function in an attached glsl file, read that function and from that, set up the sockets and inlined shader code.

That's literally what MaltFuncionNode does.

**Reply by Kolupsy** [reply]
> That's literally what MaltFuncionNode does.
Yes that is true but the MaltFunctionNode does not contain the shader function. It only stores a path which the transpiler takes to fetch the function from the different libraries. This would require me to separate shader code and custom node code. Ideally I want to keep both close together. 

> I don't get what this means. How "proper formatted glsl together with python" would look like?
I mean this sort of formatting where the variables are being replaced by the malt socket identifiers like this:
![image](https://user-images.githubusercontent.com/88381426/155874800-e2b24b23-5be1-4a21-9e2a-f0bc0764e0e5.png)

I will look into the transpiler to see if I can use it to my advantage. Maybe I should even be creating custom nodes as subclasses of the MaltFunctionNode because - as you said - thats more closely what I want in terms of design

**Comment by Kolupsy** [comment]
Update:

https://user-images.githubusercontent.com/88381426/156946493-a9847098-582f-4695-be7e-3ac0ff1f3d50.mp4

The MixRGB clone is now working. I have implemented the `get_source_code` function like this:
```py
def get_source_code(self, transpiler):
        fac = self.inputs['Factor'].get_source_initialization()
        col1 = self.inputs['Color 1'].get_source_initialization()
        col2 = self.inputs['Color 2'].get_source_initialization()
        result = self.outputs['Color'].get_source_reference()
        func = blend_functions[ self.blend_type ]
        source = '''
        vec4 {result};
        {{
        float fac = {fac};
        vec4 col1 = {col1};
        vec4 col2 = {col2};
        vec4 result;
        {func}
        {result} = result;
        }}
        '''.format(
            fac = fac,
            col1 = col1,
            col2 = col2,
            result = result,
            func = func
        )
        return source
```
I am storing all the shader code in a dictionary and only the function bodies. It looks a bit like this:
```py
blend_functions = {
    'MIX':'''
    fac = clamp(fac, 0.0, 1.0);
    result = mix( col1, col2, fac );
    result.a = col1.a;
    ''',

    'ADD':'''
    fac = clamp(fac, 0.0, 1.0);
    result = mix(col1, col1 + col2, fac);
    result.a = col1.a;
    ''',

    'MULTIPLY':'''
    fac = clamp(fac, 0.0, 1.0);
    result = mix(col1, col1 * col2, fac);
    result.a = col1.a;
    ''',

    'SUBTRACT':'''
    fac = clamp(fac, 0.0, 1.0);
    result = mix(col1, col1 - col2, fac);
    result.a = col1.a;
    ''',
...
```
The next thing I am wondering is: how can you set min, max and default values for inputs made from python nodes. I already figured out that you can set an input with `node.malt_paramters[key] = 3.1415` but I cant seem to be able to set min, max and default. I know this has to do with the new RNA system but I cant seem to figure it out.

**Comment by Kolupsy** [comment]
To generalize this, I implemented this class, that automatically sets up a function wrapper string out of the inputs and outputs of the node and inserts a function inbetween. That right now makes it pretty easy to implement the mixRGB node:
```py
class CustomFunctioNode( MaltCustomNode ):
    bl_idname = 'MaltCustomStaticNode'
    bl_label = 'Custom Static Node'
    
    def malt_setup( self ):
        super( ).malt_setup( )
        inputs = {}
        for name, type in self.get_inputs( ).items( ):
            inputs[ name ] = {'type' : type }
        outputs = {}
        for name, type in self.get_outputs( ).items( ):
            outputs[ name ] = {'type' : type }
        self.setup_sockets( inputs, outputs )
    
    def get_function_wrapper( self ):
        header = '\n'
        inputs = self.get_inputs( ).items( )
        outputs = self.get_outputs( ).items( )
        for name, type in outputs:
            var = self.outputs[name].get_source_reference( )
            header += f'{type} {var};\n'
        header += '{{\n'
        for name, type in inputs:
            var = self.inputs[ name ].get_source_initialization( )
            header += f'{type} {name} = {var};\n'
        for name, type in outputs:
            header += f'{type} {name};\n'
        header += '{func}'
        for name, type in outputs:
            var = self.outputs[name].get_source_reference( )
            header += f'{var} = {name};\n'
        header += '}}'
        return header
    
    def get_source_code( self, transpiler ):
        return self.get_function_wrapper( ).format( func = self.get_function( ))
    
    def get_inputs( self ):
        return {}
    def get_outputs( self ):
        return {}
    def get_function( self ):
        return ''
```
In the end you only need to implement `get_inputs`, `get_outputs`, `get_function` and the draw functions `draw_buttons` and `draw_label` for any subclass of the CustomFunctionNode

**Comment by Kolupsy** [comment]
@pragma37 Ok MixRGB, Math, Vector Math are done. Open quesitons at the moment are:
- [x] how to define default values, min, max and subtypes using this workflow
- [x] how to override the sockets name (subclass? in which case it would be nice to know which functions are necessary to override)
- [ ] any ideas how to substiture drivers using python
- [x] can i define a shader function inside the function block of the custom node and if so, how would i do that
- [x] how are the object attributes coming along
![image](https://user-images.githubusercontent.com/88381426/157290921-addc33bf-1cbd-4b38-aa36-6f98e326c3d2.png)
I was able dynamically hide sockets depending on the type of function used but i only set `socket.enabled = False` for example. Should I update more properties? Ive seen that `MaltSocket`s also have their own `active` attribute

**Reply by pragma37** [reply]
Looks very good!
> how to define default values, min, max and subtypes using this workflow

The inputs/outputs in setup_sockets() are (typically) the dictionaries generated by GLSL reflection. So anything that matches the META properties format works. For example, a meta property in the form of:
```glsl
/*META
   @color: subtype=Color; default=vec3(1.0, 0.1, 0.0);
*/
void foo( vec3 color)
...
```
would result in:
```python
inputs = {
  "color":  { "type" :  "vec3", "meta" : { "subtype" : "Color", "default" : (1.0, 0.1, 0.0) } }
}
```
min and max are not yet supported, but they will be added in the form of "min" and "max" metaproperties.

If you need more functionality you could also use "native" Blender properties and keep them in sync with the Malt properties by using an update callback.

> how to override the sockets name (subclass? in which case it would be nice to know which functions are necessary to override)

You can override the MaltNode.draw_socket function. But, why not just name the sockets correctly in the first place? 🤔

> any ideas how to substiture drivers using python

https://docs.blender.org/api/current/bpy.msgbus.html ?

> can i define a shader function inside the function block of the custom node and if so, how would i do that

GLSL doesn't have nested function. I guess you could declare it at the global scope:
```python
def get_source_global_parameters(self, transpiler):
    src = super().get_source_global_parameters(transpiler)
    src += """
    #ifndef FOO_DECLARED
    #define FOO_DECLARED
    void foo() {}
    #endif
    """
    return src
```

> how are the object attributes coming along

Have not looked into them yet. I'm making some updates needed for Blender 3.1.

**Reply by Kolupsy** [reply]
Thanks for the swift reply. Those are all very helpful

> You can override the MaltNode.draw_socket function. But, why not just name the sockets correctly in the first place?

I want to be able to dynamically change the displayed name and hide the type information as well. 

> GLSL doesn't have nested function. I guess you could declare it at the global scope:

cool I will try that out! When you have another function called 'foo' will that override previous definitons or throw an error?

> Have not looked into them yet. I'm making some updates needed for Blender 3.1.

oh yeah i heard it was released today. Gotta check that out

**Reply by pragma37** [reply]
> When you have another function called 'foo' will that override previous definitons or throw an error?

If the parameter types are different then it will work, otherwise it will throw an error.

**Reply by noctiswhole** [reply]
> any ideas how to substiture drivers using python

I was able to pull object info from the scene and put it into a `vec3`, but wasn't able to figure out how to setup a handler to update the `vec3` whenever the scene changes. Maybe this will help.

![image](https://user-images.githubusercontent.com/101479200/158035653-c57dc92e-72ce-44d1-96fd-660ca9d8cde5.png)


```python

import bpy
from BlenderMalt.MaltNodes.MaltNode import MaltNode


class MaltObjectInfoNode(bpy.types.Node, MaltNode):
    bl_label = "Object Info Node"

    def selected_object_update(self, context):
        # update the node tree
        print('updated')
        self.id_data.update()

    selected_object: bpy.props.StringProperty(update=selected_object_update)
    attribute_name: bpy.props.StringProperty(update=selected_object_update)

    def malt_init(self):
        self.setup()

    def malt_setup(self):
        if self.first_setup:
            self.name = 'Object Info'
        self.setup_sockets(
            {},
            {'result': {'type': 'vec3'}}
        )

    def draw_buttons(self, context, layout):
        layout.prop(self, 'attribute_name', text='Attribute')
        layout.prop_search(self, "selected_object", context.scene, "objects", text="Object")

    def get_source_socket_reference(self, socket):
        return '{}_0_{}'.format(self.get_source_name(), socket.name)

    def get_source_code(self, transpiler):
        result = self.outputs['result']

        if self.selected_object in bpy.data.objects:
            obj = bpy.data.objects[self.selected_object]
            if hasattr(obj, self.attribute_name):
                obj_attribute = getattr(obj, self.attribute_name)
                print(obj_attribute)
                initialization = 'vec3({}, {}, {})'.format(
                    obj_attribute[0],
                    obj_attribute[1],
                    obj_attribute[2],
                )
                return transpiler.declaration(result.data_type, result.array_size, result.get_source_reference(),
                                          initialization)
        initialization = 'vec3(0,0,0)'
        return transpiler.declaration(result.data_type, result.array_size, result.get_source_reference(),
                                      initialization)


classes = [
    MaltObjectInfoNode,
]


def register():
    for _class in classes: bpy.utils.register_class(_class)


def unregister():
    for _class in reversed(classes): bpy.utils.unregister_class(_class)

```

**Reply by Kolupsy** [reply]
So in order to update constantly it would need to always refresh the node tree that the node is in? But for updating I think the depsgrapgh update handlers would be the best solution

**Reply by noctiswhole** [reply]
I'm not totally sure as I'm brand new to blender plugin stuff, but I think it might be possible to store the previous value and only refresh the tree if the values change.

**Reply by pragma37** [reply]
@noctiswhole It's awesome to see other people rolling their own nodes! :)

In this case, you can avoid updating the node tree at all. Since modifying the tree triggers a shader recompilation, it's better to avoid it when possible.

You need an input socket:
```python
def malt_setup(self):
    if self.first_setup:
        self.name = 'Object Info'
    self.setup_sockets(
        {'value': {'type': 'vec3'}},
        {'result': {'type': 'vec3'}}
    )
```
In your update function, you can simply update the malt property:
```python
def selected_object_update(self, context):
    obj = bpy.data.objects[self.selected_object]
    if obj and hasattr(obj, self.attribute_name):
        self.malt_parameters['value'] = tuple(getattr(obj, self.attribute_name))
```
then, for generating the source code:
```python
def get_source_code(self, transpiler):
    result = self.outputs['result']
    value = self.inputs['value']
    return transpiler.declaration(result.data_type, result.array_size, result.get_source_reference(), value.get_source_initialization())
```

(I have not tested the code, but that's the idea)

> I was able to pull object info from the scene and put it into a vec3, but wasn't able to figure out how to setup a handler to update the vec3 whenever the scene changes. Maybe this will help.

I think you could add a `bpy.props.FloatVectorProperty` to the node, add the driver to that property, and an update function to keep the malt property in sync. (?)

**Reply by Kolupsy** [reply]
So that means that `get_source_code` gets called every time malt renders the image (so all the time)? If that is the case then yeah you could definitely have a property that dynamically updates a malt parameter. Could be interesting to try both THIS and using the depsgraph update. Im glad there might be a way to use 'driver-like' constructs on a shader level (and not just for materials)

**Reply by pragma37** [reply]
> So that means that `get_source_code` gets called every time malt renders the image (so all the time)?

Nope, only when the tree updates. In this case, the update was being forced by calling `self.id_data.update()`.

Malt parameters on the other hand **are** retrieved every time the scene has been updated. 
I'm considering adding an option to set up a python callback for malt properties, as a nicer alternative to drivers. But that could complicate optimizing BlenderMalt performance for large scenes.

**Comment by Kolupsy** [comment]
![image](https://user-images.githubusercontent.com/88381426/161290955-821af428-366a-488c-a6e3-b8c65f0ea535.png)
Little update on substituting drivers because you cannot add drivers to malt parameters. I made a test node where I have a `bpy.props.FloatProperty` with update function that updates a malt parameter (so no tree update) but the problem is that drivers and keyframes dont trigger the update function which renders this method useless. The only other thing I can think of right now is the depsgraph update handler which is on the same update cycle as drivers

**Reply by Kolupsy** [reply]
Also I am thinking about returning to the node extrusion task some time. What do you think about the little button right next to the socket knob. We could also have the socket name be the button, which would save some time but it would require every socket to have a label. I think this is the case right now but what if you want to support factor sockets in the future. Their socket layout consists of just the property at least in Eevee.

Also I am rethinking if it would not be better to split the plugin into Blender plugin and Malt plugin. I fear that the current way of inlining the shader code as a string will not scale very well

**Reply by pragma37** [reply]
I think that now that Blender has native support for node extrusion it would be better to look if it's possible to add support for custom node editors too.

> Also I am rethinking if it would not be better to split the plugin into Blender plugin and Malt plugin. I fear that the current way of inlining the shader code as a string will not scale very well

You could just include the dependency file in get_source_global_parameters:

```python
def get_source_global_parameters(self, transpiler):
    src = super().get_source_global_parameters(transpiler)
    src += f'#include "{full_path_to_the_glsl_file}"'
    return src
```

Just make sure that the glsl file has include guards.

**Comment by Kolupsy** [comment]
Custom Malt nodes work really well, I was wondering if it was possible to do the same thing for `MaltIONode` and pipeline nodes. I tried to figure out which functions I need to override to get those to work but I just could not figure it out. I know in the end you always have to have a working `get_source_code` method but especially for the socket setup I am stumped 😅

**Reply by pragma37** [reply]
I'm not sure if there's a reasonable way to wrap existing IO Nodes with a custom one. What are you trying to achieve?

On the BlenderMalt side, Pipeline nodes work pretty much the same as Shader nodes.

If you have a specific use case I may provide an example.

**Comment by Kolupsy** [comment]
> I'm not sure if there's a reasonable way to wrap existing IO Nodes with a custom one. What are you trying to achieve?

As an example I want to wrap the MAIN_PASS_PIXEL_SHADER node to be called something else, thats pretty much it but by the sound of it its probably not worth it i guess

> If you have a specific use case I may provide an example.

After a lot of fiddling around and with the help of @panthavma I was able to make a custom pipeline node (I used your bloom pass as an example to play around):

![image](https://user-images.githubusercontent.com/88381426/173638542-da5d0276-b9d9-4d6e-b91f-7da748f53280.png)

Now I am just looking at ways to write a basically custom custom node like I have done with the shader nodes before. With the benifit that I can put it into my own node category in the add menu, have control over socket visibility and node properties (enums and such) to manipulate the pipeline function.

```glsl
from Malt.PipelineNode import PipelineNode
from Malt.PipelineParameters import Parameter, Type
from Malt.GL.Texture import Texture
from Malt.GL.GL import GL_RGBA16F
from Malt.GL.RenderTarget import RenderTarget

import pathlib

_SHADER = None

BLOOMPATH = str( pathlib.Path( __file__ ).parent.parent.joinpath( 'Shaders', 'Bloom.glsl' ))

class EssentialsBloomPass( PipelineNode ):

    def __init__( self, pipeline ):
        PipelineNode.__init__( self, pipeline )
        self.resolution = None
    
    @classmethod
    def reflect_inputs( cls ):
        return {
            'Color' : Parameter( '', Type.TEXTURE ),
            'Radius' : Parameter( 0.5, Type.FLOAT ),
            'Samples' : Parameter( 64, Type.INT ),
            'Exponent' : Parameter( 2.0, Type.FLOAT ),
            'Intensity' : Parameter( 5.0, Type.FLOAT ),
        }
    
    @classmethod
    def reflect_outputs( cls ):
        return {
            'Color' : Parameter( '', Type.TEXTURE )
        }
    
    def setup_render_targets( self, resolution ):
        self.resolution = resolution
        self.texture_targets = {}
        self.texture_targets[ 'COLOR' ] = Texture( resolution, GL_RGBA16F )
        self.render_target = RenderTarget([*self.texture_targets.values( )])
    
    def execute( self, parameters ):
        inputs = parameters[ 'IN' ]
        outputs = parameters[ 'OUT' ]

        if self.pipeline.resolution != self.resolution:
            self.setup_render_targets( self.pipeline.resolution )
        
        global _SHADER
        if _SHADER is None:
            self.compile_shader( )
        
        _SHADER.textures[ 'color_texture' ] = inputs[ 'Color' ]
        _SHADER.uniforms[ 'bloom_settings' ].set_value(( inputs['Exponent'], inputs['Intensity'], inputs['Radius']))
        _SHADER.uniforms[ 'samples' ].set_value( inputs['Samples'])
        self.pipeline.common_buffer.shader_callback( _SHADER )
        self.pipeline.draw_screen_pass( _SHADER, self.render_target )
            
        outputs[ 'Color' ] = self.texture_targets[ 'COLOR' ]

    def compile_shader_from_source( self, source, include_paths = [], defines = []):
        return self.pipeline.compile_shader_from_source( source = source, include_paths = include_paths, defines = defines )

    def compile_shader( self ):
        global _SHADER
        _SHADER = self.compile_shader_from_source( f'#include "{BLOOMPATH}"' )

NODE = EssentialsBloomPass
```
Abstractly speaking I want the code above as a MaltNode subclass instead of a PipelineNode subclass.

**Reply by pragma37** [reply]
> As an example I want to wrap the MAIN_PASS_PIXEL_SHADER node to be called something else, thats pretty much it but by the sound of it its probably not worth it i guess

I think you should also take into account if changing names for the sake of it is actually a good idea, especially if you are going to distribute this as part of a plugin.
If you have feedback on the naming I think is best to just discuss it here and change it in Malt itself if needed.

> Abstractly speaking I want the code above as a MaltNode subclass instead of a PipelineNode subclass.

I think you would need a MaltNode subclass **in addition** to the MaltPipeline subclass. 
It might be possible to work around it, but the Python nodes runtime is designed to work with PipelineNodes.

If you don't want to expose the PipelineNode itself to the user, I think overriding the *reflect* method in the Pipeline like this should work:
````python
@classmethod
def reflect(cls):
    reflection = PipelineNode.reflect(cls)
    reflection['meta']['internal'] = True
    return reflection
````

Then you can just write a custom node that wraps the internal one, as you did with the shader nodes.

**Reply by Kolupsy** [reply]
> I think you would need a MaltNode subclass in addition to the MaltPipeline subclass.

That is probably a godd compromise. What I could not figure out yet was how to properly setup the sockets for pipeline nodes. I think what would really help is if the `MaltNode` class had a method called `add_socket` or something that sets up the right socket + malt parameter. Because this would allow adding sockets dynamically as well.
